src/server/worker/master.go

```diff
@@ -484,36 +503,50 @@ func (a *APIServer) waitJob(pachClient *client.APIClient, jobInfo *pps.JobInfo,
 		afterTime := startTime.Add(timeout).Sub(time.Now())
 		logger.Logf("cancelling job at: %+v", afterTime)
 		timer := time.AfterFunc(afterTime, func() {
-			if _, err := pachClient.PfsAPIClient.FinishCommit(ctx,
-				&pfs.FinishCommitRequest{
-					Commit: jobInfo.OutputCommit,
+			if jobInfo.EnableStats {
+				if _, err = pachClient.PfsAPIClient.FinishCommit(ctx, &pfs.FinishCommitRequest{
+					Commit: statsCommit,
 					Empty:  true,
 				}); err != nil {
+					logger.Logf("error from FinishCommit for stats while timing out job: %+v", err)
+				}
+			}
+			if _, err := pachClient.PfsAPIClient.FinishCommit(ctx, &pfs.FinishCommitRequest{
```

Comment:
Shouldn't this set the job state - does that already happen somewhere for this case?

Reply:
Yeah, this is one of the weird ones. This was/is piggybacking on the goroutine that is checking for the finished output commit to mark the job as killed. But, I think, after looking at it again, we probably should go ahead and update it here because that fits the job finish sequence I described in the comment and allows us to recover in the subscribe commit loop if we fail after the update job state and before the finish commit (right now we could be stuck with an incorrect job state).

Reply:
I think this change can put us in a weird situation as well if the job output commit is finished by the normal code path and the timeout occurs before we are finished (egressing, updating job state). The timeout will mark the job as killed, then fail on the finish output commit. This should ultimately be okay because the state will be updated to success by the normal code path (if the timeout update job state ran before the normal code path update job state), but there is clearly some interleaving on these concurrent code paths that can result with undesirable job state at particular points in time. What are your thoughts on how we should handle this in the short term?

Reply:
I decided to go ahead and revert because I have a better understanding of the consequences of the first model (job timeout not updating job state in the callback), which should just be cosmetic.

Reply:
We could keep the order the same, or use a local variable to note if the job was moved to a terminal state and short-circuit.